https://bugzilla.redhat.com/show_bug.cgi?id=1486223

diff --git a/gdb/nat/linux-procfs.c b/gdb/nat/linux-procfs.c
index a12f6228cb..2bf496c97b 100644
--- a/gdb/nat/linux-procfs.c
+++ b/gdb/nat/linux-procfs.c
@@ -104,7 +104,7 @@ parse_proc_status_state (const char *state)
       return PROC_STATE_TRACING_STOP;
     case 'T':
       /* Before Linux 2.6.33, tracing stop used uppercase T.  */
-      if (strcmp (state, "T (tracing stop)") == 0)
+      if (strcmp (state, "T (tracing stop)\n") == 0)
 	return PROC_STATE_TRACING_STOP;
       else
 	return PROC_STATE_STOPPED;
