--- ./gdb/dwarf2read.c	24 Aug 2008 16:39:56 -0000	1.277
+++ ./gdb/dwarf2read.c	26 Aug 2008 18:51:28 -0000
@@ -2761,7 +2769,10 @@ process_die (struct die_info *die, struc
 	 Fortran case, so we'll have to replace this gdb_assert if
 	 Fortran compilers start generating that info.  */
       processing_has_namespace_info = 1;
-      gdb_assert (die->child == NULL);
+      if (cu->language != language_fortran && die->child == NULL)
+	complaint (&symfile_complaints,
+		   _("%s at offset %d has unexpected children"),
+		   dwarf_tag_name (die->tag), die->offset);
       break;
     default:
       new_symbol (die, NULL, cu);
