https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=225783


--- gdb-6.6/gdb/testsuite/gdb.base/prelink.exp.orig	2006-03-08 19:22:09.000000000 -0500
+++ gdb-6.6/gdb/testsuite/gdb.base/prelink.exp	2007-04-08 13:22:45.000000000 -0400
@@ -48,7 +48,7 @@
     return -1
 }
 
-if {[catch "system \"prelink -qNR ${libfile}\""] != 0} {
+if {[catch "system \"/usr/sbin/prelink -qNR ${libfile}\""] != 0} {
     # Maybe we don't have prelink.
     return -1
 }
@@ -89,11 +89,11 @@
     return 0
 }
 
-if {[catch "system \"prelink -uN ${libfile}\""] != 0} {
+if {[catch "system \"/usr/sbin/prelink -uN ${libfile}\""] != 0} {
     untested "${libfile} was not prelinked, maybe system libraries are not prelinked?"
     return 0
 }
-catch "system \"prelink -qNR ${libfile}\""
+catch "system \"/usr/sbin/prelink -qNR ${libfile}\""
 
 # Start with a fresh gdb
 
