--- ./gdb/infrun.c	22 Jun 2007 12:47:48 -0000	1.243
+++ ./gdb/infrun.c	25 Jun 2007 20:43:18 -0000
@@ -466,7 +467,7 @@ static const char *scheduler_enums[] = {
   schedlock_step,
   NULL
 };
-static const char *scheduler_mode = schedlock_off;
+static const char *scheduler_mode = schedlock_step;
 static void
 show_scheduler_mode (struct ui_file *file, int from_tty,
 		     struct cmd_list_element *c, const char *value)
