--- gdb-7.4.50.20120103/gdb/configure.orig	2012-01-05 10:23:09.251299043 +0100
+++ gdb-7.4.50.20120103/gdb/configure	2012-01-05 10:23:29.921225968 +0100
@@ -682,6 +682,7 @@ LN_S
 REPORT_BUGS_TEXI
 REPORT_BUGS_TO
 PKGVERSION
+CORELOW_O
 TARGET_OBS
 subdirs
 RPM_LIBS
@@ -8504,6 +8505,7 @@ fi
 # Accumulate some settings from configure.tgt over all enabled targets
 
 TARGET_OBS=
+CORELOW_O=
 all_targets=
 
 for targ_alias in `echo $target_alias $enable_targets | sed 's/,/ /g'`
@@ -8527,12 +8529,16 @@ fi
 
     # Target-specific object files
     for i in ${gdb_target_obs}; do
-        case " $TARGET_OBS " in
-        *" ${i} "*) ;;
-        *)
-          TARGET_OBS="$TARGET_OBS ${i}"
-          ;;
-        esac
+        if test "$i" = corelow.o; then
+            CORELOW_O=corelow.o
+        else
+            case " $TARGET_OBS " in
+            *" ${i} "*) ;;
+            *)
+              TARGET_OBS="$TARGET_OBS ${i}"
+              ;;
+            esac
+        fi
     done
 
     # Check whether this target needs 64-bit CORE_ADDR
@@ -8597,10 +8603,12 @@ _ACEOF
   else
     TARGET_OBS='$(ALL_TARGET_OBS)'
   fi
+  CORELOW_O=corelow.o
 fi
 
 
 
+
 # For other settings, only the main target counts.
 gdb_sim=
 gdb_osabi=
@@ -12153,7 +12161,7 @@ for ac_header in nlist.h machine/reg.h p
 		  sys/reg.h sys/debugreg.h sys/select.h sys/syscall.h \
 		  sys/types.h sys/wait.h wait.h termios.h termio.h \
 		  sgtty.h unistd.h elf_hp.h ctype.h time.h locale.h \
-                  dlfcn.h
+		  dlfcn.h
 do :
   as_ac_Header=`$as_echo "ac_cv_header_$ac_header" | $as_tr_sh`
 ac_fn_c_check_header_mongrel "$LINENO" "$ac_header" "$as_ac_Header" "$ac_includes_default"
