From FEDORA_PATCHES Mon Sep 17 00:00:00 2001
From: Sergio Durigan Junior <sergiodj@redhat.com>
Date: Mon, 4 Mar 2019 12:08:47 -0500
Subject: gdb-fix-aarch64-enable-unittests.patch

;; Temporary fix for Aarch64 compilation when $development=false

diff --git a/gdb/gdbserver/configure.srv b/gdb/gdbserver/configure.srv
--- a/gdb/gdbserver/configure.srv
+++ b/gdb/gdbserver/configure.srv
@@ -43,7 +43,7 @@ srv_linux_obj="linux-low.o linux-osdata.o linux-procfs.o linux-ptrace.o linux-wa
 case "${target}" in
   aarch64*-*-linux*)
 			srv_regobj="arm-with-neon.o"
-			if $development; then
+			if $enable_unittests; then
 			  srv_regobj="${srv_regobj} aarch64.o"
 			  srv_regobj="${srv_regobj} linux-aarch64-tdesc-selftest.o"
                         fi
