--- gdb-6.6-orig/gdb/libunwind-frame.c	2007-04-08 20:49:09.000000000 +0200
+++ gdb-6.6/gdb/libunwind-frame.c	2007-04-10 23:45:22.000000000 +0200
@@ -74,7 +74,9 @@
 #define STRINGIFY(name)		STRINGIFY2(name)
 
 #ifndef LIBUNWIND_SO
-#define LIBUNWIND_SO "libunwind-" STRINGIFY(UNW_TARGET) ".so"
+/* Use the stable ABI major version number.  `libunwind-ia64.so' is a link time
+   only library, not a runtime one.  */
+#define LIBUNWIND_SO "libunwind-" STRINGIFY(UNW_TARGET) ".so.7"
 #endif
 
 static char *get_reg_name = STRINGIFY(UNW_OBJ(get_reg));
