https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=218379


Index: gdb-6.6/gdb/minsyms.c
===================================================================
--- gdb-6.6.orig/gdb/minsyms.c	2007-01-20 13:53:48.000000000 +0100
+++ gdb-6.6/gdb/minsyms.c	2007-01-20 13:58:47.000000000 +0100
@@ -490,6 +490,11 @@ lookup_minimal_symbol_by_pc_section (COR
 			 don't fill the bfd_section member, so don't
 			 throw away symbols on those platforms.  */
 		      && SYMBOL_BFD_SECTION (&msymbol[hi]) != NULL
+		      /* Don't ignore symbols for solib trampolines.
+		         Limit its sideeffects - only for non-0 sized trampolines.
+		         Red Hat Bug 200533 with its regression Bug 218379.  */
+		      && (MSYMBOL_TYPE (&msymbol[hi]) != mst_solib_trampoline
+		          || MSYMBOL_SIZE (&msymbol[hi]))
 		      && (!matching_bfd_sections
 			  (SYMBOL_BFD_SECTION (&msymbol[hi]), section)))
 		    {
