commit 03729380c318738d19256b219bcc03d333dc54bf
Author: Daniel Jacobowitz <dan@debian.org>
Date:   Mon Feb 1 17:51:18 2010 +0000

[...]
    	* gdb.base/interrupt.exp: Match unexpected gdb prompt.

--- a/gdb/testsuite/gdb.base/interrupt.exp
+++ b/gdb/testsuite/gdb.base/interrupt.exp
@@ -167,6 +167,7 @@ if ![file exists $binfile] then {
         # here.
 	gdb_expect {
 	    -re "^(\r\n|)data\r\n(|data\r\n)$" { pass "echo data" }
+	    -re "Undefined command.*$gdb_prompt " { fail "echo data" }
 	    timeout { fail "echo data (timeout)" }
 	    eof { fail "echo data (eof)" }
 	}
