https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=193763


Index: gdb-6.5/gdb/linespec.c
===================================================================
--- gdb-6.5.orig/gdb/linespec.c	2006-08-24 02:57:04.000000000 -0300
+++ gdb-6.5/gdb/linespec.c	2006-08-24 02:57:07.000000000 -0300
@@ -295,7 +295,7 @@ add_minsym_members (const char *class_na
      This will give us a list of all the member names including
      the function signature.  */
   completion_name = xmalloc (strlen (class_name) +
-			     strlen (member_name) + 9);
+			     strlen (member_name) + strlen("'::$base(") + 1);
   completion_name[0] = '\'';
   strcpy (completion_name+1, class_name);
   /* FIXME: make this the language class separator.  */
