--- gdb-6.8/gdb/Makefile.in-orig	2008-06-17 15:16:20.000000000 +0200
+++ gdb-6.8/gdb/Makefile.in	2008-06-17 15:17:01.000000000 +0200
@@ -1100,16 +1100,22 @@ YYOBJ = c-exp.o \
 
 DISTSTUFF = $(YYFILES)
 
+DEPDIR = .deps
+
 # Prevent Sun make from putting in the machine type.  Setting
 # TARGET_ARCH to nothing works for SunOS 3, 4.0, but not for 4.1.
 .c.o:
-	$(CC) -c $(INTERNAL_CFLAGS) $<
+	@mkdir -p $(DEPDIR)
+	$(CC) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c $(INTERNAL_CFLAGS) $<
+	@mv -f $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Po
 
 all: gdb$(EXEEXT) $(CONFIG_ALL)
 	@$(MAKE) $(FLAGS_TO_PASS) DO=all "DODIRS=`echo $(SUBDIRS) | sed 's/testsuite//'`" subdir_do
 .PHONY: all-tui
 all-tui: $(TUI)$(EXEEXT)
 
+-include $(DEPDIR)/*.Po
+
 installcheck:
 
 # The check target can not use subdir_do, because subdir_do does not
