https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=218379


--- gdb-6.5-depatched/gdb/minsyms.c	2006-12-17 16:10:53.000000000 -0500
+++ gdb-6.5/gdb/minsyms.c	2006-12-17 16:51:21.000000000 -0500
@@ -505,6 +505,11 @@
 			  don't fill the bfd_section member, so don't
 			  throw away symbols on those platforms.  */
 		       && SYMBOL_BFD_SECTION (&msymbol[hi]) != NULL
+		      /* Don't ignore symbols for solib trampolines.
+		         Limit its sideeffects - only for non-0 sized trampolines.
+		         Red Hat Bug 200533 with its regression Bug 218379.  */
+		      && (MSYMBOL_TYPE (&msymbol[hi]) != mst_solib_trampoline
+		          || MSYMBOL_SIZE (&msymbol[hi]))
 		      && (!matching_bfd_sections
 			  (SYMBOL_BFD_SECTION (&msymbol[hi]), section)))
 		  --hi;
