https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=200048


diff -ru gdb-6.5-orig/gdb/tui/tui-layout.c gdb-6.5-pr200048/gdb/tui/tui-layout.c
--- gdb-6.5-orig/gdb/tui/tui-layout.c	2006-08-07 21:33:42.000000000 +0200
+++ gdb-6.5-pr200048/gdb/tui/tui-layout.c	2006-08-13 02:26:39.000000000 +0200
@@ -519,13 +519,17 @@
     {
     case SRC_COMMAND:
     case SRC_DATA_COMMAND:
-      find_line_pc (cursal.symtab,
+      if (find_line_pc (cursal.symtab,
 		    TUI_SRC_WIN->detail.source_info.start_line_or_addr.u.line_no,
-		    &pc_list, &num_pc_values);
-      /* FIXME: What do we do with multiple pc values for ctors/dtors or
-                inlined functions?  */
-      addr = pc_list[0];
-      xfree (pc_list);
+		    &pc_list, &num_pc_values))
+	{
+	  /* FIXME: What do we do with multiple pc values for ctors/dtors or
+		    inlined functions?  */
+	  addr = pc_list[0];
+	  xfree (pc_list);
+	}
+      else
+        addr = 0;
       break;
     case DISASSEM_COMMAND:
     case SRC_DISASSEM_COMMAND:
